// rsuite/esm/Notification/Notification.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _taggedTemplateLiteralLoose from "@babel/runtime/helpers/esm/taggedTemplateLiteralLoose";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/esm/objectWithoutPropertiesLoose";

var _templateObject, _templateObject2;

import React, { useCallback, useState, useContext } from 'react';
import PropTypes from 'prop-types';
import { useClassNames, useTimeout, MESSAGE_STATUS_ICONS, useIsMounted } from '../utils';
import CloseButton from '../CloseButton';
import ToastContext from '../toaster/ToastContext';
var Notification = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var _props$as = props.as,
      Component = _props$as === void 0 ? 'div' : _props$as,
      _props$classPrefix = props.classPrefix,
      classPrefix = _props$classPrefix === void 0 ? 'notification' : _props$classPrefix,
      closable = props.closable,
      _props$duration = props.duration,
      duration = _props$duration === void 0 ? 4500 : _props$duration,
      className = props.className,
      type = props.type,
      header = props.header,
      children = props.children,
      onClose = props.onClose,
      rest = _objectWithoutPropertiesLoose(props, ["as", "classPrefix", "closable", "duration", "className", "type", "header", "children", "onClose"]);

  var _useState = useState('show'),
      display = _useState[0],
      setDisplay = _useState[1];

  var _useClassNames = useClassNames(classPrefix),
      withClassPrefix = _useClassNames.withClassPrefix,
      merge = _useClassNames.merge,
      prefix = _useClassNames.prefix;

  var isMounted = useIsMounted();

  var _useContext = useContext(ToastContext),
      usedToaster = _useContext.usedToaster; // Timed close message


  var _useTimeout = useTimeout(onClose, duration, usedToaster && duration > 0),
      clear = _useTimeout.clear; // Click to trigger to close the message


  var handleClose = useCallback(function (event) {
    setDisplay('hiding');
    onClose === null || onClose === void 0 ? void 0 : onClose(event);
    clear();
    setTimeout(function () {
      if (isMounted()) {
        setDisplay('hide');
      }
    }, 1000);
  }, [onClose, clear, isMounted]);
  var renderHeader = useCallback(function () {
    if (!header) {
      return null;
    }

    return /*#__PURE__*/React.createElement("div", {
      className: prefix('title')
    }, type ? /*#__PURE__*/React.createElement("div", {
      className: prefix(_templateObject || (_templateObject = _taggedTemplateLiteralLoose(["title-with-icon"])))
    }, MESSAGE_STATUS_ICONS[type], header) : /*#__PURE__*/React.createElement("div", {
      className: prefix('title')
    }, header));
  }, [header, type, prefix]);

  if (display === 'hide') {
    return null;
  }

  var classes = merge(className, withClassPrefix(type, display, {
    closable: closable
  }));
  return /*#__PURE__*/React.createElement(Component, _extends({
    role: "alert"
  }, rest, {
    ref: ref,
    className: classes
  }), /*#__PURE__*/React.createElement("div", {
    className: prefix(_templateObject2 || (_templateObject2 = _taggedTemplateLiteralLoose(["content"])))
  }, renderHeader(), /*#__PURE__*/React.createElement("div", {
    className: prefix('description')
  }, typeof children === 'function' ? children() : children)), closable && /*#__PURE__*/React.createElement(CloseButton, {
    onClick: handleClose
  }));
});
Notification.displayName = 'Notification';
Notification.propTypes = {
  as: PropTypes.elementType,
  duration: PropTypes.number,
  header: PropTypes.node,
  closable: PropTypes.bool,
  classPrefix: PropTypes.string,
  className: PropTypes.string,
  type: PropTypes.oneOf(['info', 'success', 'warning', 'error']),
  onClose: PropTypes.func
};
export default Notification;
