// rsuite/cjs/Dropdown/Dropdown.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _omit = _interopRequireDefault(require("lodash/omit"));

var _pick = _interopRequireDefault(require("lodash/pick"));

var _DropdownMenu = _interopRequireDefault(require("./DropdownMenu"));

var _utils = require("../utils");

var _deprecatePropType = _interopRequireDefault(require("../utils/deprecatePropType"));

var _DropdownItem = _interopRequireDefault(require("./DropdownItem"));

var _DropdownContext = _interopRequireDefault(require("./DropdownContext"));

var _Menu = _interopRequireDefault(require("../Menu/Menu"));

var _DropdownToggle = _interopRequireDefault(require("./DropdownToggle"));

var _kebabCase = _interopRequireDefault(require("lodash/kebabCase"));

var _NavContext = _interopRequireDefault(require("../Nav/NavContext"));

var _DropdownState = require("./DropdownState");

var _warnOnce = _interopRequireDefault(require("../utils/warnOnce"));

var _Nav = _interopRequireDefault(require("../Nav"));

var _DropdownSeparator = _interopRequireDefault(require("./DropdownSeparator"));

/**
 * The <Dropdown> API
 * When used inside <Sidenav>, renders a <TreeviewRootItem>;
 * Otherwise renders a <MenuRoot>
 */
var Dropdown = /*#__PURE__*/_react.default.forwardRef(function (props, ref) {
  var activeKey = props.activeKey,
      onSelect = props.onSelect,
      rest = (0, _objectWithoutPropertiesLoose2.default)(props, ["activeKey", "onSelect"]);
  var _rest$as = rest.as,
      Component = _rest$as === void 0 ? 'div' : _rest$as,
      title = rest.title,
      onClose = rest.onClose,
      onOpen = rest.onOpen,
      onToggle = rest.onToggle,
      _rest$trigger = rest.trigger,
      trigger = _rest$trigger === void 0 ? 'click' : _rest$trigger,
      _rest$placement = rest.placement,
      placement = _rest$placement === void 0 ? 'bottomStart' : _rest$placement,
      toggleAs = rest.toggleAs,
      toggleClassName = rest.toggleClassName,
      open = rest.open,
      defaultOpen = rest.defaultOpen,
      _rest$classPrefix = rest.classPrefix,
      classPrefix = _rest$classPrefix === void 0 ? 'dropdown' : _rest$classPrefix,
      className = rest.className,
      disabled = rest.disabled,
      children = rest.children,
      menuStyle = rest.menuStyle,
      style = rest.style,
      toggleProps = (0, _objectWithoutPropertiesLoose2.default)(rest, ["as", "title", "onClose", "onOpen", "onToggle", "trigger", "placement", "toggleAs", "toggleClassName", "open", "defaultOpen", "classPrefix", "className", "disabled", "children", "menuStyle", "style"]);
  var nav = (0, _react.useContext)(_NavContext.default);

  var _useClassNames = (0, _utils.useClassNames)(classPrefix),
      merge = _useClassNames.merge,
      withClassPrefix = _useClassNames.withClassPrefix;

  var _useClassNames2 = (0, _utils.useClassNames)('dropdown-menu'),
      withMenuClassPrefix = _useClassNames2.withClassPrefix,
      mergeMenuClassName = _useClassNames2.merge;

  var menuButtonTriggers = (0, _react.useMemo)(function () {
    if (!trigger) {
      return undefined;
    }

    var triggerMap = {
      hover: 'mouseover',
      click: 'click',
      contextMenu: 'contextmenu'
    };

    if (!Array.isArray(trigger)) {
      return [triggerMap[trigger]];
    }

    return trigger.map(function (t) {
      return triggerMap[t];
    });
  }, [trigger]);

  var _useReducer = (0, _react.useReducer)(_DropdownState.reducer, _DropdownState.initialState),
      items = _useReducer[0].items,
      dispatch = _useReducer[1];

  var hasSelectedItem = (0, _react.useMemo)(function () {
    return items.some(function (item) {
      return item.props.selected;
    });
  }, [items]);
  var dropdownContextValue = (0, _react.useMemo)(function () {
    return {
      activeKey: activeKey,
      onSelect: onSelect,
      hasSelectedItem: hasSelectedItem,
      dispatch: dispatch
    };
  }, [activeKey, onSelect, hasSelectedItem, dispatch]); // Deprecate <Dropdown> within <Nav> usage
  // in favor of <Nav.Menu> API

  if (nav) {
    (0, _warnOnce.default)('Usage of <Dropdown> within <Nav> is deprecated. Replace with <Nav.Menu>');
    return /*#__PURE__*/_react.default.createElement(_Nav.default.Menu, (0, _extends2.default)({
      ref: ref
    }, props));
  }

  var renderMenuButton = function renderMenuButton(menuButtonProps, menuButtonRef) {
    return /*#__PURE__*/_react.default.createElement(_DropdownToggle.default, (0, _extends2.default)({
      ref: menuButtonRef,
      as: toggleAs,
      className: toggleClassName,
      placement: placement,
      disabled: disabled
    }, (0, _omit.default)(menuButtonProps, ['open']), (0, _omit.default)(toggleProps, ['data-testid'])), title);
  };

  return /*#__PURE__*/_react.default.createElement(_DropdownContext.default.Provider, {
    value: dropdownContextValue
  }, /*#__PURE__*/_react.default.createElement(_Menu.default, {
    open: open,
    defaultOpen: defaultOpen,
    menuButtonText: title,
    renderMenuButton: renderMenuButton,
    disabled: disabled,
    openMenuOn: menuButtonTriggers,
    renderMenuPopup: function renderMenuPopup(_ref, popupRef) {
      var open = _ref.open,
          popupProps = (0, _objectWithoutPropertiesLoose2.default)(_ref, ["open"]);
      var menuClassName = mergeMenuClassName(className, withMenuClassPrefix({}));
      return /*#__PURE__*/_react.default.createElement("ul", (0, _extends2.default)({
        ref: popupRef,
        className: menuClassName,
        style: menuStyle,
        hidden: !open
      }, popupProps), children);
    },
    onToggleMenu: function onToggleMenu(open) {
      onToggle === null || onToggle === void 0 ? void 0 : onToggle(open);

      if (open) {
        onOpen === null || onOpen === void 0 ? void 0 : onOpen();
      } else {
        onClose === null || onClose === void 0 ? void 0 : onClose();
      }
    }
  }, function (_ref2, menuContainerRef) {
    var _withClassPrefix;

    var open = _ref2.open,
        menuContainer = (0, _objectWithoutPropertiesLoose2.default)(_ref2, ["open"]);
    var classes = merge(className, withClassPrefix((_withClassPrefix = {}, _withClassPrefix["placement-" + (0, _kebabCase.default)((0, _utils.placementPolyfill)(placement))] = !!placement, _withClassPrefix.disabled = disabled, _withClassPrefix.open = open, _withClassPrefix['selected-within'] = hasSelectedItem, _withClassPrefix)));
    return /*#__PURE__*/_react.default.createElement(Component, (0, _extends2.default)({
      ref: (0, _utils.mergeRefs)(ref, menuContainerRef),
      className: classes
    }, menuContainer, (0, _pick.default)(toggleProps, ['data-testid']), {
      style: style
    }));
  }));
});

Dropdown.Item = _DropdownItem.default;
Dropdown.Menu = _DropdownMenu.default;
Dropdown.Separator = _DropdownSeparator.default;
Dropdown.displayName = 'Dropdown';
Dropdown.propTypes = {
  activeKey: _propTypes.default.any,
  classPrefix: _propTypes.default.string,
  trigger: _propTypes.default.oneOfType([_propTypes.default.array, _propTypes.default.oneOf(['click', 'hover', 'contextMenu'])]),
  placement: _propTypes.default.oneOf(_utils.PLACEMENT_8),
  title: _propTypes.default.node,
  disabled: _propTypes.default.bool,
  icon: _propTypes.default.node,
  menuStyle: _propTypes.default.object,
  className: _propTypes.default.string,
  toggleClassName: _propTypes.default.string,
  children: _propTypes.default.node,
  open: (0, _deprecatePropType.default)(_propTypes.default.bool),
  eventKey: _propTypes.default.any,
  as: _propTypes.default.elementType,
  toggleAs: _propTypes.default.elementType,
  noCaret: _propTypes.default.bool,
  style: _propTypes.default.object,
  onClose: _propTypes.default.func,
  onOpen: _propTypes.default.func,
  onToggle: _propTypes.default.func,
  onSelect: _propTypes.default.func,
  onMouseEnter: _propTypes.default.func,
  onMouseLeave: _propTypes.default.func,
  onContextMenu: _propTypes.default.func,
  onClick: _propTypes.default.func,
  renderToggle: _propTypes.default.func
};
var _default = Dropdown;
exports.default = _default;
