// rsuite-table/lib/Column.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = exports.columnHandledProps = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

// eslint-disable-next-line @typescript-eslint/no-unused-vars
function Column(_props) {
  return null;
}

var propTypes = {
  align: _propTypes["default"].oneOf(['left', 'center', 'right']),
  verticalAlign: _propTypes["default"].oneOf(['top', 'middle', 'bottom']),
  width: _propTypes["default"].number,
  fixed: _propTypes["default"].oneOfType([_propTypes["default"].bool, _propTypes["default"].oneOf(['left', 'right'])]),
  resizable: _propTypes["default"].bool,
  sortable: _propTypes["default"].bool,
  flexGrow: _propTypes["default"].number,
  minWidth: _propTypes["default"].number,
  colSpan: _propTypes["default"].number,
  rowSpan: _propTypes["default"].func,
  treeCol: _propTypes["default"].bool,
  onResize: _propTypes["default"].func,
  children: _propTypes["default"].node,
  fullText: _propTypes["default"].bool
};
Column.displayName = 'Table.Column';
Column.defaultProps = {
  width: 100
};
Column.propTypes = propTypes;
var columnHandledProps = Object.keys(propTypes);
exports.columnHandledProps = columnHandledProps;
var _default = Column;
exports["default"] = _default;
