// rsuite/esm/Message/Message.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _taggedTemplateLiteralLoose from "@babel/runtime/helpers/esm/taggedTemplateLiteralLoose";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/esm/objectWithoutPropertiesLoose";

var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5;

import React, { useState, useCallback, useContext } from 'react';
import PropTypes from 'prop-types';
import { useClassNames, useTimeout, MESSAGE_STATUS_ICONS, STATUS, useIsMounted } from '../utils';
import CloseButton from '../CloseButton';
import ToastContext from '../toaster/ToastContext';
var Message = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var _withClassPrefix;

  var _props$as = props.as,
      Component = _props$as === void 0 ? 'div' : _props$as,
      className = props.className,
      _props$classPrefix = props.classPrefix,
      classPrefix = _props$classPrefix === void 0 ? 'message' : _props$classPrefix,
      children = props.children,
      closable = props.closable,
      _props$duration = props.duration,
      duration = _props$duration === void 0 ? 2000 : _props$duration,
      full = props.full,
      header = props.header,
      _props$type = props.type,
      type = _props$type === void 0 ? 'info' : _props$type,
      showIcon = props.showIcon,
      onClose = props.onClose,
      rest = _objectWithoutPropertiesLoose(props, ["as", "className", "classPrefix", "children", "closable", "duration", "full", "header", "type", "showIcon", "onClose"]);

  var _useState = useState('show'),
      display = _useState[0],
      setDisplay = _useState[1];

  var _useClassNames = useClassNames(classPrefix),
      withClassPrefix = _useClassNames.withClassPrefix,
      merge = _useClassNames.merge,
      prefix = _useClassNames.prefix;

  var isMounted = useIsMounted();

  var _useContext = useContext(ToastContext),
      usedToaster = _useContext.usedToaster; // Timed close message


  var _useTimeout = useTimeout(onClose, duration, usedToaster && duration > 0),
      clear = _useTimeout.clear;

  var handleClose = useCallback(function (event) {
    setDisplay('hiding');
    onClose === null || onClose === void 0 ? void 0 : onClose(event);
    clear();
    setTimeout(function () {
      if (isMounted()) {
        setDisplay('hide');
      }
    }, 1000);
  }, [clear, isMounted, onClose]);

  if (display === 'hide') {
    return null;
  }

  var classes = merge(className, withClassPrefix(type, display, (_withClassPrefix = {
    full: full
  }, _withClassPrefix['has-title'] = header, _withClassPrefix['has-icon'] = showIcon, _withClassPrefix)));
  return /*#__PURE__*/React.createElement(Component, _extends({
    role: "alert"
  }, rest, {
    ref: ref,
    className: classes
  }), /*#__PURE__*/React.createElement("div", {
    className: prefix(_templateObject || (_templateObject = _taggedTemplateLiteralLoose(["container"])))
  }, closable && /*#__PURE__*/React.createElement(CloseButton, {
    onClick: handleClose
  }), showIcon && /*#__PURE__*/React.createElement("div", {
    className: prefix(_templateObject2 || (_templateObject2 = _taggedTemplateLiteralLoose(["icon-wrapper"])))
  }, MESSAGE_STATUS_ICONS[type]), /*#__PURE__*/React.createElement("div", {
    className: prefix(_templateObject3 || (_templateObject3 = _taggedTemplateLiteralLoose(["content"])))
  }, header && /*#__PURE__*/React.createElement("div", {
    className: prefix(_templateObject4 || (_templateObject4 = _taggedTemplateLiteralLoose(["header"])))
  }, header), children && /*#__PURE__*/React.createElement("div", {
    className: prefix(_templateObject5 || (_templateObject5 = _taggedTemplateLiteralLoose(["body"])))
  }, children))));
});
Message.displayName = 'Message';
Message.propTypes = {
  type: PropTypes.oneOf(STATUS),
  className: PropTypes.string,
  onClose: PropTypes.func,
  closable: PropTypes.bool,
  title: PropTypes.node,
  description: PropTypes.node,
  showIcon: PropTypes.bool,
  full: PropTypes.bool,
  classPrefix: PropTypes.string
};
export default Message;
